pkg/bgpv1/agent/controller.go

```diff
@@ -195,6 +214,132 @@ func (c *Controller) Run(ctx context.Context) {
 	}
 }
 
+// Reconcile is the main reconciliation loop for the BGP Control Plane Controller.
+// It is responsible for determining the current mode of BGP control plane, which can be disabled, bgpv1 or bgpv2.
+// Based on presence of BGP peering policy and BGP node config, it will apply the appropriate configuration.
+// Following is the state transition table for the controller:
+// Initial state         | BGPPP exists | BGPNC exists | Action	                    | Next state
+// ----------------------|--------------|--------------|----------------------------|-----------
+// disabled              | true         | don't care   | Apply BGPv1                | bgpv1
+// disabled              | false        | true         | Apply BGPv2                | bgpv2
+// disabled              | false        | false        | Do nothing                 | disabled
+// bgpv1                 | true         | don't care   | Apply BGPv1                | bgpv1
+// bgpv1                 | false        | true         | Delete BGPv1, Apply BGPv2  | bgpv2
+// bgpv1                 | false        | false        | Delete BGPv1               | disabled
+// bgpv2                 | true         | don't care   | Delete BGPv2, Apply BGPv1  | bgpv1
+// bgpv2                 | false        | true         | Apply BGPv2                | bgpv2
+// bgpv2                 | false        | false        | Delete BGPv2               | disabled
+func (c *Controller) Reconcile(ctx context.Context) error {
+	bgpp, err := c.bgppSelection()
+	if err != nil {
+		log.WithError(err).Error("bgp peering policy selection failed")
+		return err
+	}
+	bgppExists := bgpp != nil
+
+	bgpnc, bgpncExists, err := c.BGPNodeConfigStore.GetByKey(resource.Key{
+		Name: c.LocalCiliumNode.Name,
+	})
+	if err != nil {
+		log.WithError(err).Error("failed to get BGPNodeConfig")
+		return err
+	}
+
+	switch c.Mode {
+	case Disabled:
+		if bgppExists {
+			err = c.reconcileBGPP(ctx, bgpp)
+		} else if bgpncExists {
+			err = c.reconcileBGPNC(ctx, bgpnc)
+		}
+
+	case BGPv1:
+		if bgppExists {
+			err = c.reconcileBGPP(ctx, bgpp)
+		} else {
+			c.cleanupBGPP(ctx)
+
+			// check if we need to reconcile bgpv2
+			if bgpncExists {
+				err = c.reconcileBGPNC(ctx, bgpnc)
+			}
+		}
+
+	case BGPv2:
+		if bgppExists {
+			// delete bgpv2 and apply bgpv1
+			c.cleanupBGPNC(ctx)
```

Comment:
Is it safe to proceed with `reconcileBGPP` if `cleanupBGPNC` failed? (similarly in the `BGPv1` case).

Reply:
I am not entirely sure, previously we had [best effort cleanup](https://github.com/cilium/cilium/pull/32426/files#diff-663c8aba9492de183ebe045c8bf425d6a6a645970cd2a201a7073e9b21ffe8adL326) as well. 

When cleanup fails, it is mostly because of a bug. And most likely next creation is going to fail because of left overs like bind on port 179.

Reply:
Looking at the code, I don't see where cleanup can fail.

Reply:
I see, good that we at least log it now in case of something completely unexpected.

Reply:
Yep.